src/core_atmosphere/physics/physics_wrf/module_sf_sfcdiags_ruclsm.F

```diff
@@ -0,0 +1,222 @@
+          qsfcprox = qlev1+qfx(I,J)/(rho*cqs(I,J))
+          qsfcmr = qsfc(i,j)/(1.-qsfc(i,j))
+
+          ! print *,'qsfc,qsfcmr,qsfcprox,qlev1',qsfc(i,j),qsfcmr,qsfcprox,qlev1
+          ! print *,'(qsfcprox-qsfcmr)/qsfcmr =', (qsfcprox-qsfcmr)/qsfcmr
+
+          if ( flux ) then
+            if(cqs2(I,J).lt.1.E-5) then
+            ! - under very stable conditions use first level for 2-m mixing ratio
+              q2(I,J)=qlev1
+            else
+              x2m = qsfcprox - qfx(i,j)/(rho*cqs2(i,j))
+              q2(i,j) = x2m
+            endif
+          else
+          ! QFX is not used
+            q2(I,J) = qsfcmr - cqs(I,J)/cqs2(i,j)*(qsfcmr - qlev1)
+          endif  ! flux
+
+          !    print *,'FINAL - qsfc,qsfcmr,qsfcprox,q2(i,j),qlev1', &
+          !                     qsfc(i,j),qsfcmr,qsfcprox,q2(i,j),qlev1
+          !    print *,'(q2-qlev1)/qlev1 =', (q2(i,j)-qlev1)/qlev1
+
+! Alternative logarithmic diagnostics:
+          dt = t3d(i,1,j) - tsk(i,j)
+          dq = qlev1 - qsfcmr
+          dz1= 0.5 * dz(i,1,j)
+          if (dt > 0.) then
+          ! stable stratification        
+             fh  = min(max(1.-dt/10.,0.01), 1.0)
+             !for now, set zt = 0.05
+             fac = log((2.  + .05)/(0.05 + fh))/ &
+                   log((dz1 + .05)/(0.05 + fh))
+             t2_alt(i,j) = tsk(i,j) + fac*(t3d(i,1,j) - tsk(i,j))
+          else
+             !no alternatives (yet) for unstable conditions
+             t2_alt(i,j) = t2(i,j)
+          endif
+          if (dq > 0.) then
+             fh  = min(max(1.-dq/0.003,0.01), 1.0)
+             !for now, set zt = 0.05
+             fac = log((2.  + .05)/(0.05 + fh))/ &
+                   log((dz1 + .05)/(0.05 + fh))
+             q2_alt(i,j) = qsfcmr + fac*(qlev1 - qsfcmr)
+          else
+             !no alternatives (yet) for unstable conditions
+             q2_alt(i,j) = q2(i,j)
+          endif
+
+! Use alternative diagnostics for stable stratification
+          t2(i,j) = t2_alt(i,j)
+          q2(i,j) = q2_alt(i,j)
+
+          ! check that T2 values lie in the range between TSK and T at the 1st level
+          x2m     = MAX(MIN(tsk(i,j),t3d(i,1,j)) , t2(i,j))
+          t2(i,j) = MIN(MAX(tsk(i,j),t3d(i,1,j)) , x2m)
+          th2(i,j) = t2(i,j)*(1.e5/psfc)**rovcp
+
+          ! check that q2 values lie between qsfcmr and qlev1
+          x2m     = MAX(MIN(qsfcmr,qlev1) , q2(i,j))
+          q2(i,j) = MIN(MAX(qsfcmr,qlev1) , x2m)
+          ! saturation check
+          tempc = t2(i,j)-273.15
+          if (tempc .le. 0.0) then
+          ! ice and supercooled water
+            qsat = rsif(psfc, t2(i,j))
+          else
+          ! water
+            qsat = rslf(psfc, t2(i,j))
+          endif
+          q2(i,j) = min(qsat, q2(i,j))
+
+        enddo
+      enddo
+
+  end subroutine sfcdiags_ruclsm
+
+!- saturation functions are from Thompson microphysics scheme
```

Comment:
I will put these functions in a utilities module so they can imported and used. This is fine for now though.